### PR TITLE
Support data validation using value and tolerance (subclass implementation)

### PR DESCRIPTION
**Pull request by phackstock**

### CODE REVIEW

File: nomenclature/processor/data_validator.py [hunk 1]
```diff
@@ -54,6 +74,19 @@ class DataValidator(Processor):
...
+    @field_validator("criteria_items", mode="before")
+    def check_criteria(cls, v):
+        for criterion in v:
+            bounds = ["upper_bound", "lower_bound"]
+            values = ["value", "atol", "rtol"]
+            if any(bound in criterion for bound in bounds) and any(
+                value in criterion for value in values
+            ):
+                raise ValueError(
+                    f"Cannot mix bounds and value criteria. Found: {criterion}"
```
Comment (danielhuppmann) [inline nomenclature/processor/data_validator.py:86]:
```suggestion
                    f"Cannot use bounds and value-criteria simultaneously: {criterion}"
```

File: tests/test_validate_data.py [hunk 2]
```diff
@@ -34,8 +34,8 @@ def test_DataValidator_from_file():
     "name, match",
     [
         ("missing_criteria", "No validation criteria provided. Found"),
-        ("bounds_and_value", "Cannot use value and bounds simultaneously. Found"),
-        ("bounds_and_rtol", "Cannot use tolerance with bounds. Use"),
+        ("bounds_and_value", "Cannot mix bounds and value criteria. Found:"),
+        ("bounds_and_rtol", "Cannot mix bounds and value criteria. Found:"),
```
Comment (danielhuppmann) [inline tests/test_validate_data.py:38]:
```suggestion
        ("bounds_and_value", "Cannot use bounds and value-criteria simultaneously:"),
        ("bounds_and_rtol", "Cannot use bounds and value-criteria simultaneously:"),
```

File: nomenclature/processor/data_validator.py [hunk 3]
```diff
@@ -54,6 +74,19 @@ class DataValidator(Processor):
...
     file: Path
 
+    @field_validator("criteria_items", mode="before")
+    def check_criteria(cls, v):
+        for criterion in v:
+            bounds = ["upper_bound", "lower_bound"]
+            values = ["value", "atol", "rtol"]
+            if any(bound in criterion for bound in bounds) and any(
+                value in criterion for value in values
+            ):
```
Comment (danielhuppmann) [inline nomenclature/processor/data_validator.py:84]: A proposal to make this more readable...
```suggestion
            has_bounds = any(c in criterion for c in ["upper_bound", "lower_bound"])
            has_values = any(c in criterion for c in ["value", "atol", "rtol"])
            if has_bounds and has_values:
```

File: nomenclature/processor/data_validator.py [hunk 4]
```diff
@@ -1,50 +1,70 @@
...
 
-    @model_validator(mode="before")
-    def check_validation_criteria_exist(cls, values):
-        if values.get("upper_bound") is None and values.get("lower_bound") is None:
+    @model_validator(mode="after")
+    def check_validation_criteria_exist(self):
+        if self.upper_bound is None and self.lower_bound is None:
             raise ValueError(
-                "No validation criteria provided. Found " + str(cls.criteria)
+                "No validation criteria provided. Found " + str(self.criteria)
```
Comment (danielhuppmann) [inline nomenclature/processor/data_validator.py:62]:
```suggestion
                "No validation criteria provided: " + str(self.criteria)
```

File: tests/test_validate_data.py [hunk 5]
```diff
@@ -34,8 +34,8 @@ def test_DataValidator_from_file():
     "name, match",
     [
         ("missing_criteria", "No validation criteria provided. Found"),
```
Comment (danielhuppmann) [inline tests/test_validate_data.py:36]:
```suggestion
        ("missing_criteria", "No validation criteria provided:"),
```

**Review by danielhuppmann — APPROVED** [review]
Nice work! Increasing the pydantery to the maximum... 

A few suggestions inline, then good to be merged!

File: nomenclature/processor/data_validator.py (hunk 3 above)
Comment (phackstock) [reply nomenclature/processor/data_validator.py:84]: Ah very nice, that one's much better 👍

**Pull request merged**